Added docs related with usage (#604)
kubearchive/kubearchive: docs/modules/ROOT/pages/getting-started/configuration.adoc

====

The pod keeps existing on Kubernetes, occupying space:

[source,bash]
----
$ kubectl get pods --namespace default
NAME     READY   STATUS      RESTARTS   AGE
fedora   0/1     Completed   0          4m50s
----

== `deleteWhen`: Deleting Resources

They key feature of KubeArchive is deleting resources. This feature keeps
the Kubernetes cluster free of resources that are not needed anymore. To enable
deletion of resources use the `deleteWhen` key.

[IMPORTANT]
====
When resources are deleted with `deleteWhen` they get archived too.
====

The following KubeArchiveConfig configures KubeArchive to delete (and archive)
pods when they match `status.phase == "Succeeded"`:

[source,yaml]
----
---
apiVersion: kubearchive.kubearchive.org/v1alpha1
kind: KubeArchiveConfig
metadata:
  name: kubearchive
  namespace: default
spec:
  resources:
    - selector:
        apiVersion: v1
        kind: Pod
      deleteWhen: status.phase == "Succeeded"
----

To see it in action apply the KubeArchiveConfig in your namespace and
run a pod:

[source,bash]
----
kubectl run auto-deleted --image quay.io/fedora/fedora:latest --restart Never -- echo "sleep 10"
----

> Start watching pods to see how this pod gets created, executed, completed and deleted:

Reviewer: ```suggestion
Start watching pods to see how this pod is created, executed, completed and deleted:
```

----
---
apiVersion: kubearchive.kubearchive.org/v1alpha1
kind: KubeArchiveConfig
metadata:
  name: kubearchive
  namespace: default
spec:
  resources:
    - selector:
        apiVersion: v1
        kind: Pod
      archiveOnDelete: status.phase == "Succeeded"
----

To see it in action apply the KubeArchiveConfig in your namespace and
run a couple of pods:

[source,bash]
----
kubectl run failed --image quay.io/fedora/fedora:latest --restart Never -- false
kubectl run archived-on-deletion --image quay.io/fedora/fedora:latest --restart Never -- echo "hello world"
----

Wait for them to fail and complete and then delete them:

[source,bash]
----
kubectl delete pod archived-on-deletion
kubectl delete pod failed
----

Query KubeArchvie to check that only the pod that completed correctly (`archived-on-deletion`)
was archived:

[source,bash]
----
$ curl --insecure \
    -H "Authorization: Bearer ${SA_TOKEN}" \
    https://localhost:8081/api/v1/namespaces/default/pods \
    | jq -r '.items[] | [.metadata.name, .metadata.uid] | @csv'

...
"archived-on-deletion","2c5fd5f6-cdab-4d6b-b008-b3f5cff5df9e"
----

== Next Steps

These are the three main functionalities KubeArchive offers related

> to resource archiving. Explore this documentation to learn more

Reviewer: ```suggestion
to resource archiving. Explore the documentation to learn more
```

= Configuring KubeArchive

This document explains how to configure KubeArchive so you can archive 
resources and query them.

== Prerequisites
* KubeArchive installed and running in a cluster (xref:getting-started/installation.adoc[])
* Know how to query the KubeArchive API (xref:getting-started/kubearchive-api.adoc[])

== The KubeArchiveConfig resource

> To configure KubeArchive to archive or delete resources create a

Reviewer 2: Even though we use `deleteWhen` as key for removing resources from etcd I think we should establish a difference between "delete a resource" (meaning, removing it forever) and "remove from the kubernetes cluster storage or etcd". We currently don't provide a feature for "deleting resources from KubeArchive database" but maybe we are in the future so I'd like to be more explicit about this.

If I'm not mistaken "delete" usually imply unrecoverable while "remove" isn't that extreme meaning "take away" or "set aside" so I would suggest we use "remove" in our docs to prevent confussion.

```suggestion
To configure KubeArchive to archive or remove resources from the cluster create a
```

Reviewer 3: I tihnk we should either use "delete" or change to "removeWhen".

Reviewer 3: And I actually vote to keep delete. We are in fact deleting it from the cluster. It no longer exists there.

> We currently don't provide a feature for "deleting resources from KubeArchive database" but maybe we are in the future so I'd like to be more explicit about this.

I view this as database work which is outside the scope of KubeArchive. Cleaning data from the database is not something I think KubeArchive should ever do.

Reviewer 2: > I tihnk we should either use "delete" or change to "removeWhen".

I don't think that both things are related, the resource is being "deleted" from the cluster (in fact, the command is `kubectl delete`) but the resource isn't deleted, but removed if considered the big picture because it's kind of "moved" from etcd to an external place.

> I view this as database work which is outside the scope of KubeArchive. Cleaning data from the database is not something I think KubeArchive should ever do.

Even though if we don't consider having this feature in the scope of KubeArchive, I still find "deletion" a hard word considering that the resources are still available to query.

If you think that using the "delete" word isn't going to cause any missunderstanding, I'm ok with leaving it, but I thought that the "softness" of the "remove" word is more clear with the concept.

Author: I agree with keeping delete. I can add the "from the cluster" to make it more specific.

Reviewer 2: > However, this is all just semantics which we can debate all day. :-)

Right, and that's why I'm open to keep the "delete" even if the "remove" sounds better for me, but it's ok, at the end we all mean the same idea.

> I can add the "from the cluster" to make it more specific.

This is fine.

>  You can't see them using kubectl or any other client that provides access to the cluster.

Be careful with this statement, as the cli provided by KubeArchive is provided as a `kubectl` plugin, meaning you can actually retrieve the resources with `kubectl` if the plugin is installed.

Feel free to close these threads @rh-hemartin :)

Reviewer 3: > Be careful with this statement, as the cli provided by KubeArchive is provided as a kubectl plugin, meaning you can actually retrieve the resources with kubectl if the plugin is installed.

This is a great point. I forgot about that. But if I install the plugin then I would say I already know about KubeArchive.

KubeArchiveConfig custom resource. KubeArchiveConfigs are limited to 1 per namespace
and should be named `kubearchive`. KubeArchiveConfigs have this general form:

[source,yaml]
----
---
apiVersion: kubearchive.kubearchive.org/v1alpha1
kind: KubeArchiveConfig
metadata:
  name: kubearchive
  namespace: default
spec:
  resources: [...]
----

`spec.resources` is a list of elements, each defining rules for a specific kind, so KubeArchive knows

> when to archive or delete them.

Reviewer 2: Based on the previous suggestion:
```suggestion
when to archive or remove them.
```

> spec:
>   resources: [...]
> ----
> 
> `spec.resources` is a list of elements, each defining rules for a specific kind, so KubeArchive knows
> when to archive or delete them.

Reviewer 2: This is a good place to use the [callouts feature](https://docs.asciidoctor.org/asciidoc/latest/verbatim/callouts/#callout-syntax) from asciidocs.

```suggestion
spec:
  resources: [...] <1>
----
<1> `spec.resources` is a list of elements, each defining rules for a specific kind, so KubeArchive knows
when to archive or delete them.
```

> ==  `selector`: Selecting Resources

Reviewer 3: I don't think we should document this at all, but rather point to the `ApiServerSource` documentation for this.

Otherwise we are duplicating things and have to keep them in sync.

Author: I think pointing them to Knative's documentation will be more confusing as they will assume we support Knative's ApiServerSource `selector`, which allows for `matchLabels`, `matchExpression`... However we don't use that when we process the KubeArchiveConfig and create an ApiServerSource. We are only using the `kind` and `apiVersion` (func `parseConfigMap` in the operator code).

Resources in the KubeArchiveConfig are defined using a `selector` which
requires two keys: `kind` and `apiVersion`. Each entry on `spec.resources`
requires a `selector`:

[source,yaml]
----
---
apiVersion: kubearchive.kubearchive.org/v1alpha1
kind: KubeArchiveConfig
metadata:
  name: kubearchive
  namespace: default
spec:
  resources:
    - selector:
        apiVersion: v1
        kind: Pod
      ...
    - selector:
        apiVersion: batch/v1
        kind: Job
      ...
    - selector:
        apiVersion: apps/v1
        kind: Deployment
      ...
----

[TIP]
====

> The `kind` provided in a selector must be singular. For example `kind: Pod` is valid, but `kind: Pods` is invalid.

Reviewer 2: If we specify this we probably want to clarify also that the word should be capitalized as it is based on the `kind` property of the resource definition and not on the `singular` or `plural` name properties.

```suggestion
The `kind` property in a selector should not be confused with the `singular` or `plural` properties of the resource. 
It should be in singular and capitalized. 
For example `kind: Pod` is valid, but `kind: Pods` is invalid.
```

Author: Hmm.. I think we are getting confusing with this block. I fundamentally agree with the changes you propose Marta, and after reading them I also think this block shouldn't exist. `kind` is `kind` and I feel that specifying what it or it isn't makes it more confusing. I will remove this block and if users start wondering what `kind` is, then we can bring it back.

====

With each entry on `spec.resources` KubeArchive requires one of
`archiveWhen`, `deleteWhen` and `archiveOnDelete`. These keys accept
a string which is an expression in the
link:https://cel.dev[CEL language format].
The expressions are evaluated every time a resource defined by `selector`

> changes or gets deleted. The expressions must evaluate to either true or false.

Reviewer 3: I might be good to mention that `true` is a valid expression to always have something happen.

`status.phase == "Succeeded"` condition:

[source,yaml]
----
---
apiVersion: kubearchive.kubearchive.org/v1alpha1
kind: KubeArchiveConfig
metadata:
  name: kubearchive
  namespace: default
spec:
  resources:
    - selector:
        apiVersion: v1
        kind: Pod
      archiveWhen: status.phase == "Succeeded"
----

To see it in action apply the KubeArchiveConfig in your namespace and
run a pod:

[source,bash]
----
kubectl run fedora --image quay.io/fedora/fedora:latest --restart Never -- sleep 10
----

The pod will not appear in KubeArchive until it completes. After completion, it is present in KubeArchive:

[source,bash]
----
$ curl --insecure \
    -H "Authorization: Bearer ${SA_TOKEN}" \
    https://localhost:8081/api/v1/namespaces/default/pods \
    | jq -r '.items[] | [.metadata.name, .metadata.uid] | @csv'

"fedora", "a3bdb6d2-b683-4913-9d24-e01af60c94e3"
----

[NOTE]
====
We use `jq` in these examples to reduce output.
====

The pod remains in Kubernetes, occupying space:

[source,bash]
----
$ kubectl get pods --namespace default
NAME     READY   STATUS      RESTARTS   AGE
fedora   0/1     Completed   0          4m50s
----

== `deleteWhen`: Deleting Resources

> They key feature of KubeArchive is deleting resources. This feature keeps

Reviewer 2: Typo and remove vs delete:
```suggestion
The key feature of KubeArchive is removing resources from the Kubernetes cluster. This feature keeps
```

the Kubernetes cluster free of resources that are not needed anymore. To enable

> deletion of resources use the `deleteWhen` key.

Reviewer 2: Based on previous suggestions:
```suggestion
the removal of resources use the `deleteWhen` key.
```

[IMPORTANT]
====

> When resources are deleted with `deleteWhen` they get archived too.

Reviewer 2: Based on the previous comment:
```suggestion
When resources are removed with `deleteWhen` they get archived too.
```

====

> The following KubeArchiveConfig configures KubeArchive to delete (and archive)

Reviewer 2: Based on the previous suggestion:
```suggestion
The following KubeArchiveConfig configures KubeArchive to remove (and archive)
```

pods when they match `status.phase == "Succeeded"`:

[source,yaml]
----
---
apiVersion: kubearchive.kubearchive.org/v1alpha1
kind: KubeArchiveConfig
metadata:
  name: kubearchive
  namespace: default
spec:
  resources:
    - selector:
        apiVersion: v1
        kind: Pod
      deleteWhen: status.phase == "Succeeded"
----

To see it in action apply the KubeArchiveConfig in your namespace and
run a pod:

[source,bash]
----
kubectl run auto-deleted --image quay.io/fedora/fedora:latest --restart Never -- echo "sleep 10"
----

> Start watching pods to see how this pod is created, executed, completed and deleted:

Reviewer 2: Based on the previous suggestion:
```suggestion
Start watching pods to see how this pod is created, executed, completed and removed:
```

[source,bash]
----
$ kubectl get pods -w
NAME        READY   STATUS              RESTARTS   AGE
auto-deleted   0/1     ContainerCreating   0          2s
auto-deleted   1/1     Running             0          2s
auto-deleted   0/1     Completed           0          13s
auto-deleted   0/1     Completed           0          14s
auto-deleted   0/1     Terminating         0          14s
auto-deleted   0/1     Terminating         0          14s
----

After the pod has been deleted from the cluster by KubeArchive, retrieve it:

[source,bash]
----
$ curl --insecure \
    -H "Authorization: Bearer ${SA_TOKEN}" \
    https://localhost:8081/api/v1/namespaces/default/pods \
    | jq -r '.items[] | [.metadata.name, .metadata.uid] | @csv'

...
"auto-deleted","64c48176-ba8c-4f2a-a662-1fd660f7a3b6"
----

> == `archiveOnDelete`: Archiving on Deletion

Reviewer 2: Based o n the previous suggestion:

```suggestion
== `archiveOnDelete`: Archiving on Deletion from the cluster
```

I kept the "delete" verb in here to match the `archiveOnDelete` key but added the context of "from the cluster" to be more explicit of the scope of that deletion.

KubeArchive also can be used with existing applications that clean up resources.
This enables you to keep using a specialized tool for deletion and use KubeArchive
to store the resources. The following KubeArchiveConfig configures KubeArchive

> to archive pods when they get deleted only if they match the condition `status.phase == "Succeeded"`

Reviewer 2: Based on the previous suggestion:
```suggestion
to archive pods when they get removed from the cluster only if they match the condition `status.phase == "Succeeded"`
```

so failed pods that get deleted do not get archived.

[source,yaml]
----
---
apiVersion: kubearchive.kubearchive.org/v1alpha1
kind: KubeArchiveConfig
metadata:
  name: kubearchive
  namespace: default
spec:
  resources:
    - selector:
        apiVersion: v1
        kind: Pod
      archiveOnDelete: status.phase == "Succeeded"
----

To see it in action apply the KubeArchiveConfig in your namespace and
run a couple of pods:

[source,bash]
----
kubectl run failed --image quay.io/fedora/fedora:latest --restart Never -- false
kubectl run archived-on-deletion --image quay.io/fedora/fedora:latest --restart Never -- echo "hello world"
----

> Wait for them to fail and complete and then delete them:

Reviewer 2: Based on the previous suggestion:
```suggestion
Wait for them to fail and complete and then remove them:
```

[source,bash]
----
kubectl delete pod archived-on-deletion
kubectl delete pod failed
----

> Query KubeArchvie to check that only the pod that completed correctly (`archived-on-deletion`)

Reviewer 2: Typo:
```suggestion
Query KubeArchive to check that only the pod that completed correctly (`archived-on-deletion`)
```
